Source/WebCore/dom/InternalObserverInspect.cpp

```diff
@@ -0,0 +1,252 @@
+            auto scope = DECLARE_CATCH_SCOPE(vm);
+
+            m_inspector.error->handleEventRethrowingException(value);
+
+            JSC::Exception* exception = scope.exception();
+            if (UNLIKELY(exception)) {
+                scope.clearException();
+                protectedSubscriber()->error(exception->value());
+                return;
+            }
+        }
+
+        protectedSubscriber()->error(value);
+    }
+
+    void complete() final
+    {
+        InternalObserver::complete();
+
+        removeAbortHandler();
+
+        if (m_inspector.complete) {
+            Ref vm = this->protectedVM();
+            JSC::JSLockHolder lock(vm);
+            auto scope = DECLARE_CATCH_SCOPE(vm);
+
+            m_inspector.complete->handleEventRethrowingException();
+
+            JSC::Exception* exception = scope.exception();
+            if (UNLIKELY(exception)) {
+                scope.clearException();
+                protectedSubscriber()->error(exception->value());
+                return;
+            }
+        }
+
+        protectedSubscriber()->complete();
+    }
+
+    template<typename VisitorType>
+    void visitAdditionalChildrenInternal(VisitorType& visitor) const
+    {
+        protectedSubscriber()->visitAdditionalChildren(visitor);
+        if (m_inspector.next)
+            m_inspector.next->visitJSFunction(visitor);
+        if (m_inspector.error)
+            m_inspector.error->visitJSFunction(visitor);
+        if (m_inspector.complete)
+            m_inspector.complete->visitJSFunction(visitor);
+        if (m_inspector.subscribe)
+            m_inspector.subscribe->visitJSFunction(visitor);
+        if (m_inspector.abort)
+            m_inspector.abort->visitJSFunction(visitor);
+    }
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor& visitor) const final
+    {
+        visitAdditionalChildrenInternal(visitor);
+    }
+
+    void visitAdditionalChildren(JSC::SlotVisitor& visitor) const final
+    {
+        visitAdditionalChildrenInternal(visitor);
+    }
+
+    void removeAbortHandler()
+    {
+        if (!m_abortAlgorithmHandler)
+            return;
+
+        auto handle = std::exchange(m_abortAlgorithmHandler, std::nullopt);
+        protectedSubscriber()->protectedSignal()->removeAlgorithm(*handle);
+    }
+
+    Ref<JSC::VM> protectedVM() const
+    {
+        auto* globalObject = protectedScriptExecutionContext()->globalObject();
+        ASSERT(globalObject);
+        return globalObject->vm();
+    }
+
+    Ref<Subscriber> protectedSubscriber() const
+    {
+        return m_subscriber;
+    }
+
+    InternalObserverInspect(ScriptExecutionContext& context, Ref<Subscriber>&& subscriber, ObservableInspector&& inspector)
+        : InternalObserver(context)
+        , m_subscriber(WTFMove(subscriber))
+        , m_inspector(WTFMove(inspector))
+    {
+        if (m_inspector.abort) {
+            Ref signal = protectedSubscriber()->protectedSignal();
```

Comment:
`Ref signal = protectedSubscriber()->signal();`

`protectedFoo()` functions are only used to avoid local variables, not to initialize them.